drivers/FlashIAP.cpp

```diff
@@ -86,21 +94,39 @@ int FlashIAP::program(const void *buffer, uint32_t addr, uint32_t size)
 {
     uint32_t page_size = get_page_size();
     uint32_t current_sector_size = flash_get_sector_size(&_flash, addr);
-    // addr and size should be aligned to page size, and multiple of page size
-    // page program should not cross sector boundaries
-    if (!is_aligned(addr, page_size) ||
-        !is_aligned(size, page_size) ||
-        (size < page_size) ||
-        (((addr % current_sector_size) + size) > current_sector_size)) {
+    uint32_t chunk, prog_size;
+    const uint8_t *buf = (uint8_t *) buffer;
+
+    // addr should be aligned to page size
+    if (!is_aligned(addr, page_size) || (!buffer)) {
```

Comment:
Do we not still need to check that size is aligned and that program does not go past end of the device?

Reply:
The page alignment check here is for the address only. Don't think we need the size alignment check. Adding such a check would require the user to split the writes to the aligned and unaligned ones, which is basically what the new code does.
Regarding the device end check - good thought. Will add that.